File: core/src/main/scala/mybatis/scala/support/typing/typehandler/DelegateScalaOptionTypeHandler.scala
package mybatis.scala.support.typing.typehandler

import org.apache.ibatis.`type`.JdbcType

import java.sql.{CallableStatement, PreparedStatement, ResultSet}

class DelegateScalaOptionTypeHandler[T](delegate: TypeHandler[T]) extends TypeHandler[scala.Option[T]]:

  override def setParameter(ps: PreparedStatement, i: Int, parameter: Option[T], jdbcType: JdbcType): Unit =
    parameter match
      case Some(value) => delegate.setParameter(ps, i, value, jdbcType)
      case None        => delegate.setParameter(ps, i, null.asInstanceOf[T], jdbcType)

  override def getResult(rs: ResultSet, columnName: String): Option[T] =
    Option(delegate.getResult(rs, columnName))

  override def getResult(rs: ResultSet, columnIndex: Int): Option[T] =
    Option(delegate.getResult(rs, columnIndex))

  override def getResult(cs: CallableStatement, columnIndex: Int): Option[T] =
    Option(delegate.getResult(cs, columnIndex))

object DelegateScalaOptionTypeHandler extends DelegateScalaOptionTypeHandlerInstances

trait DelegateScalaOptionTypeHandlerInstances:
  given [T](using delegate: TypeHandler[T]): TypeHandler[scala.Option[T]] =
    new DelegateScalaOptionTypeHandler[T](delegate)

object DelegateScalaOptionTypeHandlerInstances extends DelegateScalaOptionTypeHandlerInstances

File: core/src/main/java/mybatis/scala/support/typing/DelegateScalaOptionTypeHandler.java
package mybatis.scala.support.typing;

import org.apache.ibatis.type.JdbcType;
import org.apache.ibatis.type.TypeHandler;
import scala.Option;

import java.sql.CallableStatement;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public abstract class DelegateScalaOptionTypeHandler<T> implements TypeHandler<Option<T>> {

    private final TypeHandler<T> typeHandler;

    protected DelegateScalaOptionTypeHandler(TypeHandler<T> typeHandler) {
        this.typeHandler = typeHandler;
    }

    @Override
    public void setParameter(PreparedStatement ps, int i, Option<T> parameter, JdbcType jdbcType) throws SQLException {
        if (parameter.isDefined()) {
            typeHandler.setParameter(ps, i, parameter.get(), jdbcType);
        } else {
            typeHandler.setParameter(ps, i, null, jdbcType);
        }
    }

    @Override
    public Option<T> getResult(ResultSet rs, String columnName) throws SQLException {
        T result = typeHandler.getResult(rs, columnName);
        return Option.apply(result);
    }

    @Override
    public Option<T> getResult(ResultSet rs, int columnIndex) throws SQLException {
        T result = typeHandler.getResult(rs, columnIndex);
        return Option.apply(result);
    }

    @Override
    public Option<T> getResult(CallableStatement cs, int columnIndex) throws SQLException {
        T result = typeHandler.getResult(cs, columnIndex);
        return Option.apply(result);
    }

}
